README.md
## About Laravel PayFast Onsite
![GitHub release (latest by date)](https://img.shields.io/github/v/release/fintech-systems/laravel-payfast-onsite) [![Build Status](https://app.travis-ci.com/fintech-systems/laravel-payfast-onsite.svg?branch=main)](https://app.travis-ci.com/fintech-systems/laravel-payfast-onsite) ![GitHub](https://img.shields.io/github/license/fintech-systems/laravel-payfast-onsite)

A [PayFast Onsite Payments](https://developers.payfast.co.za/docs#onsite_payments) implementation for Laravel designed to ease subscription billing. [Livewire](https://laravel-livewire.com/) views are included.

**THIS IS BETA SOFTWARE**

- There may be some bugs but the core functionality should work.

Requirements:

- PHP 8.0
- Laravel
- A [PayFast account](https://www.payfast.co.za/registration)

## Installation

Install the package via composer:

```bash
composer require fintech-systems/laravel-payfast-onsite
```

## Publish Configuration and Views

Publish the config file with:
```bash
php artisan vendor:publish --provider="FintechSystems\Payfast\PayfastServiceProvider" --tag="payfast-config"
```

Publish the Success and Cancelled views and the Livewire components for subscriptions and receipts.

```bash
php artisan vendor:publish --provider="FintechSystems\Payfast\PayfastServiceProvider" --tag="payfast-views"
```

### Nova Integration

Optionally publish Laravel Nova Subscription and Receipts Resources

```bash
php artisan vendor:publish --provider="FintechSystems\Payfast\PayfastServiceProvider" --tag="payfast-nova-resource"
```

## Migrations

A migration is needed to create Customers, Orders, Receipts and Subscriptions tables:

```bash
php artisan migrate
```

## Example Configuration

`config/payfast.php`:

```php
<?php

return [
    'merchant_id' => env('PAYFAST_MERCHANT_ID', '10004002'),
    'merchant_key' => env('PAYFAST_MERCHANT_KEY', 'q1cd2rdny4a53'),
    'passphrase' => env('PAYFAST_PASSPHRASE', 'payfast'),
    'testmode' => env('PAYFAST_TESTMODE', true),        
    'return_url' => env('PAYFAST_RETURN_URL', config('app.url') . '/payfast/success'),
    'cancel_url' => env('PAYFAST_CANCEL_URL', config('app.url') . '/payfast/cancel'),
    'notify_url' => env('PAYFAST_NOTIFY_URL', config('app.url') . '/payfast/webhook'),
    'card_update_link_css' => env('CARD_UPDATE_LINK_CSS', 'inline-flex items-center px-4 py-2 bg-gray-800 border border-transparent rounded-md font-semibold text-xs text-white uppercase tracking-widest hover:bg-gray-700 active:bg-gray-900 focus:outline-none focus:border-gray-900 focus:ring focus:ring-gray-300 disabled:opacity-25 transition'),
    'card_updated_return_url' => env('CARD_UPDATED_RETURN_URL', config('app.url') . '/user/profile'),
    'plans' => [
        3 => [
            'name' => 'Monthly R 99',
            'start_date' => \Carbon\Carbon::now()->addDay()->format('Y-m-d'),
            'payfast_frequency' => 3,
            'initial_amount' => 5.99,
            'recurring_amount' => 5.99,
        ],
        6 => [
            'name' => 'Yearly R 1089',
            'start_date' => \Carbon\Carbon::now()->format('Y-m-d'),
            'payfast_frequency' => 6,
            'initial_amount' => 6.89,
            'recurring_amount' => 6.89,
        ]
    ],
    'cancelation_reasons' => [
        'Too expensive',
        'Lacks features',
        'Not what I expected',
    ],
];
```

## Livewire Setup

### Jetstream

I have modelled some Livewire views to fit into a [Laravel Jetstream](https://jetstream.laravel.com) user profile page.

When calling the Livewire component, you can override any [PayFast form field](https://developers.payfast.co.za/docs#step_1_form_fields) by specifying a `mergeFields` array.

Example modification Jetstream Livewire's `resources/views/profiles/show.php`:

```php
<!-- Subscriptions -->
<div class="mt-10 sm:mt-0">    
    @livewire('jetstream-subscriptions', ['mergeFields' => [
            'name_first' => $user->name,
            'name_last' => $user->name,
            'item_description' => 'Subscription to Online Service'
        ]] )        
</div>

<x-jet-section-border />
<!-- End Subscriptions -->

<!-- Receipts -->
    <div class="mt-10 sm:mt-0">
        @livewire('jetstream-receipts')
    </div>

<x-jet-section-border />
<!-- End Receipts -->
```

## Usage

### Examples

- Generate a payment link
- Create an ad-hoc token optionally specifying the amount
- Cancel a subscription
- Update a card

```php
use FintechSystems\Payfast\Facades\Payfast;

Route::get('/payment', function() {
    return Payfast::payment(5,'Order #1');
});

Route::get('/cancel-subscription', function() {
    return Payfast::cancelSubscription('73d2a218-695e-4bb5-9f62-383e53bef68f');
});

Route::get('/create-subscription', function() {
    return Payfast::createSubscription(
        Carbon::now()->addDay()->format('Y-m-d'),
        5, // Amount
        6 // Frequency (6 = annual, 3 = monthly)
    );
});

Route::get('/create-adhoc-token', function() {
    return Payfast::createAdhocToken(5);
});

Route::get('/fetch-subscription', function() {
    return Payfast::fetchSubscription('21189d52-12eb-4108-9c0e-53343c7ac692');
});

Route::get('/update-card', function() {
    return Payfast::updateCardLink('40ab3194-20f0-4814-8c89-4d2a6b5462ed');
});
```

## Testing

```bash
vendor/bin/phpunit
```

## Changelog

Please see [CHANGELOG](CHANGELOG.md) for more information on what has changed recently.

## Screenshots

![Livewire Subscriptions and Receipts Components](../../blob/main/screenshots/subscription_and_receipts.png)

## Credits

- [Eugene van der Merwe](https://github.com/eugenevdm)

## License

The MIT License (MIT). Please see [License File](LICENSE.md) for more information.


CHANGELOG.md
# Changelog

All notable changes to `laravel-payfast-onsite` will be documented in this file.

## 0.8.0 - 2022-05-30

- Webhook will now store the subscription using the word "default"
- Completely redid the Livewire blade and used boilerplate as reference

## 0.7.2 - 2022-05-30

- Simplified the way the Livewire blade determines subscription status

## 0.7.0 - 2022-05-30

- Change signature for createOnsitePayments to accept merge fields and updated README.md

## 0.6.3 - 2022-05-30

- Readme updates
- Fixed more composer stuff after rename. Abandoning `laravel-payfast` on Packagist in favour of `laravel-payfast-onsite`

## 0.6.2 - 2022-05-30

- More fixes after rename this time in service provider

## 0.6.1 - 2022-05-30

- Fix namespacing after rename

## 0.6.0 - 2022-05-30

- Dropped the word 'Payfast' from the Livewire components
- Renamed repo to laravel-payfast-onsite
- Added to the README to specify a better description and also added `config/payfast.php`

## 0.5.5 - 2022-05-29

- Added an update card method in the Livewire component
- Added a helper method in the service provider to easily identify users in grace period
- Added a method to easily deduce current active subscription
- Work on dates when user cancels current subscription, still have to fix amount

## 0.5.4 - 2022-05-29

- Added PHP code fixer
- Configured Travis
- Removed two currency locale tests because Travis failed
- Removed a lot of old paddle code or renamed it, especially paddleOptions
- There is a now a Receipt Nova interface and a Livewire front-end for receipts too
- Added documentation references to `TROUBLESHOOTING.md`
- Had to comment out four tests as API testing not working yet
- Payment method was added to the subscriptions and receipts table

## 0.5.3 - 2022-05-28

- Added more information about local debugging especially with regards to changing the `.env` for Expose / ngrok
- Added Nova assets, a Subscription resource and two actions, Fetch Subscription and Cancel Subscription
- The Livewire Jetstream blade should now interpret if you are in grace period (e.g. you have cancelled)
- Added code to get the PayFast subscription `status_text` and `run_date` post Subscription creation so that the UI can show when the next billing will occur
- Worked on spacing of lines and logging and commenting
- Fixed some links to live code in README

## 0.5.2 - 2022-05-27

- Added information about setup to README and new TROUBLESHOOTING file
- Created alpha release

## 0.5.1 - 2021-12-26

- Breaking change Subscription table now has all fields returned by Fetch Subscription from PayFast API
- Added payment_method to subscriptions table
- Updated composer
- A new test to check if subscription payments can be handled
- A basic receipt test was copied from Cashier Paddle
- An *incomplete* Fetch subscription test was created
- Fetch subscription now returns the array from PayFast
- A new PayFast API service provider was created that implements the code in the constructor
- There is a new method to update a PayFast subscription
- Fairly close to a beta candidate as most of the database schema work is done
- Tests *failing*

## 0.4.1 - 2021-12-14

- migrated payfast-jetstream blade back into this library
- add $user in Subscription module and removed 'message' listener
- started implementation of trials
- the cancel dialog will now close on cancellation
- removed some debugging code, both in the Subscription module and the Subscription concern
- updated the webhook to handle card information updates
- added IP address to order information and updates many tests
- added merchant_payment_info to many tests as it caused nullable errors
- all tests passing

## 0.4.0 - 2021-12-12

- added ip_address to orders table, breaking change
- order ID doesn't have to be unique anymore in receipts table
- don't send return value when receiving an exception in the webhook
- reverse engineered onsite.engine.js and obtained a working event listener 'message'
- added event listeners for post PayFast subscription modal event 'message'

## 0.3.0 - 2021-12-12

- did a lot of work on the Jetstream/Livewire component to get dynamic subscritions working
- breaking change is createOnsitePayment signature change

## 0.2.0 - 2021-12-10

- the system can now handle ad-hoc payments and will store the receipts
- refactored database to handle ad-hoc receipts
- created boilerplate to fetch subscription information after appying a payment
- refactored the webhook to make it look a lot more tidy
- move another routine into try / catch block
- told the bank to get off and fix my card which doesn't work - will get a a Thyme bank card tomorrow
- improved logging sentence with full stop and more ray colour coding in webhook

## 0.1.0 - 2021-12-07

- more refactoring to keep payfast class small and compact
- first attempt at getting onsite payments work - card locked out with 'Unfortunately this card does not support recurring payments.'
- the order is now a numeric and a dash with today's date and time
- improvement on invalid morph model exception message

## 0.0.3 - 2021-12-07

- refactored webhooks
- all tests passing for the first time

## 0.0.2 - 2021-12-05

- ability to create a new ad-hoc subscription with any amount
- ability to cancel a subscription
- copied many of the Laravel Paddle tests and made them work
- foundation jetstream block to show subscription information
- ability to create a new ad-hoc subscription with R 0 amount
- pay button generation

## 0.0.1 - 2021-12-03

- initial release


config/payfast.php
<?php

return [
    'merchant_id' => env('PAYFAST_MERCHANT_ID', '10004002'),
    'merchant_key' => env('PAYFAST_MERCHANT_KEY', 'q1cd2rdny4a53'),
    'passphrase' => env('PAYFAST_PASSPHRASE', 'payfast'),
    'testmode' => env('PAYFAST_TESTMODE', true),
    'return_url' => env('PAYFAST_RETURN_URL', config('app.url') . '/payfast/success'),
    'cancel_url' => env('PAYFAST_CANCEL_URL', config('app.url') . '/payfast/cancel'),
    'notify_url' => env('PAYFAST_NOTIFY_URL', config('app.url') . '/payfast/webhook'),
    'card_update_link_css' => env('CARD_UPDATE_LINK_CSS', 'inline-flex items-center px-4 py-2 bg-gray-800 border border-transparent rounded-md font-semibold text-xs text-white uppercase tracking-widest hover:bg-gray-700 active:bg-gray-900 focus:outline-none focus:border-gray-900 focus:ring focus:ring-gray-300 disabled:opacity-25 transition'),
    'card_updated_return_url' => env('CARD_UPDATED_RETURN_URL', config('app.url') . '/user/profile'),
    'plans' => [
        3 => [
            'name' => 'Monthly R 99',
            'start_date' => \Carbon\Carbon::now()->format('Y-m-d'),
            'payfast_frequency' => 3, // 3 = monthly
            'initial_amount' => 5.99, // For card updates or reactivatitions, this should be zero
            'recurring_amount' => 5.99,
        ],
        6 => [
            'name' => 'Yearly R 1089',
            'start_date' => \Carbon\Carbon::now()->format('Y-m-d'),
            'payfast_frequency' => 6, // 6 = yearly
            'initial_amount' => 6.89, // For card updates or reactivatitions, this should be zero
            'recurring_amount' => 6.89,
        ]
    ],
    'cancelation_reasons' => [
        'Too expensive',
        'Lacks features',
        'Not what I expected',
    ],
];
